File: apis/elysium-chat-api/src/main/java/com/seventh_root/elysium/api/chat/ChatChannelPipelineComponent.java
package com.seventh_root.elysium.api.chat;

import com.seventh_root.elysium.chat.exception.ChatChannelMessageFormattingFailureException;

public abstract class ChatChannelPipelineComponent implements Comparable<ChatChannelPipelineComponent> {

    public enum Type {

        PRE_PROCESSOR(0),
        FORMATTER(1),
        POST_PROCESSOR(2);

        private int priority;

        Type(int priority) {
            this.priority = priority;
        }

        public int getPriority() {
            return priority;
        }

    }

    @Override
    public int compareTo(ChatChannelPipelineComponent component) {
        return getType().getPriority() - component.getType().getPriority();
    }

    public abstract Type getType();

    public abstract String process(String message, ChatMessageContext context) throws ChatChannelMessageFormattingFailureException;

}

File: apis/elysium-chat-api/src/main/kotlin/com/seventh_root/elysium/api/chat/ChatChannelPipelineComponent.kt
package com.seventh_root.elysium.api.chat

import com.seventh_root.elysium.api.chat.exception.ChatChannelMessageFormattingFailureException

abstract class ChatChannelPipelineComponent : Comparable<ChatChannelPipelineComponent> {

    enum class Type private constructor(val priority: Int) {

        PRE_PROCESSOR(0),
        FORMATTER(1),
        POST_PROCESSOR(2)

    }

    override fun compareTo(other: ChatChannelPipelineComponent): Int {
        return type.priority - other.type.priority
    }

    abstract val type: Type

    @Throws(ChatChannelMessageFormattingFailureException::class)
    abstract fun process(message: String, context: ChatMessageContext): String?

    @Throws(ChatChannelMessageFormattingFailureException::class)
    abstract fun postProcess(message: String, context: ChatMessagePostProcessContext): String?

}
